File: packages/filters-gl/src/bevelFilter.ts
import type { GlRenderTarget } from '@flighthq/render-gl';
import { clearGlRenderTarget } from '@flighthq/render-gl';
import type { BevelFilter } from '@flighthq/types';
import type { GlRenderState } from '@flighthq/types';

import { applyBoxBlurFilterToGl } from './blurFilter';
import { applyGlBlitOffsetPass, applyGlBlitPass, applyGlTintPass } from './tintShader';

/**
 * Applies a bevel filter to `source`, writing the result to `dest`.
 *
 * Compositing order:
 *   1. Shadow layer (blurred alpha at +offset, shadow color)
 *   2. Highlight layer (blurred alpha at -offset, highlight color)
 *   3. Source on top (unless `knockout`)
 *
 * `bevelType` controls which layers are drawn:
 *   - `'full'` (default): both shadow and highlight
 *   - `'outer'`: shadow and highlight placed outside the source boundary
 *   - `'inner'`: shadow and highlight placed inside the source boundary
 *
 * `scratch` must contain three render targets of the same dimensions as `dest`.
 * The filter allocates nothing itself.
 */
export function applyBevelFilterToGl(
  state: GlRenderState,
  source: GlRenderTarget,
  dest: GlRenderTarget,
  scratch: GlRenderTarget[],
  filter: Readonly<Omit<BevelFilter, 'type'>>,
): void {
  const angle = ((filter.angle ?? 45) * Math.PI) / 180;
  const distance = filter.distance ?? 4;
  const dx = Math.cos(angle) * distance;
  const dy = Math.sin(angle) * distance;
  const shadowColor = filter.shadowColor ?? 0x000000;
  const shadowAlpha = filter.shadowAlpha ?? 1;
  const highlightColor = filter.highlightColor ?? 0xffffff;
  const highlightAlpha = filter.highlightAlpha ?? 1;
  const strength = filter.strength ?? 1;
  const quality = Math.max(1, Math.round(filter.quality ?? 1));
  const knockout = filter.knockout ?? false;
  const bevelType = filter.bevelType ?? 'full';

  const [tinted, blurred, blurTemp] = scratch;

  // Build the shared blur basis (neutral white tint preserves alpha shape)
  applyGlTintPass(state, source, tinted, 0xffffff, 1, strength);
  applyBoxBlurFilterToGl(state, tinted, blurred, blurTemp, {
    blurX: filter.blurX ?? 4,
    blurY: filter.blurY ?? 4,
    passes: quality,
  });

  clearGlRenderTarget(state, dest);

  if (bevelType === 'full' || bevelType === 'outer') {
    // Shadow at +offset
    applyGlTintPass(state, blurred, tinted, shadowColor, shadowAlpha, 1);
    applyGlBlitOffsetPass(state, tinted, dest, dx, dy);
    // Highlight at -offset
    applyGlTintPass(state, blurred, tinted, highlightColor, highlightAlpha, 1);
    applyGlBlitOffsetPass(state, tinted, dest, -dx, -dy);
  } else {
    // inner: shadow at -offset (opposite direction puts it inside), highlight at +offset
    applyGlTintPass(state, blurred, tinted, shadowColor, shadowAlpha, 1);
    applyGlBlitOffsetPass(state, tinted, dest, -dx, -dy);
    applyGlTintPass(state, blurred, tinted, highlightColor, highlightAlpha, 1);
    applyGlBlitOffsetPass(state, tinted, dest, dx, dy);
  }

  if (!knockout) {
    applyGlBlitPass(state, source, dest);
  }
}

File: crates/flighthq-filters-gl/src/bevel_filter.rs
//! Bevel GL filter pass.
//!
//! Compositing order:
//!   1. Shadow layer (blurred alpha at +offset, shadow color)
//!   2. Highlight layer (blurred alpha at -offset, highlight color)
//!   3. Source on top (unless `filter.knockout`)
//!
//! `bevel_type` controls which layers are drawn:
//!   - `Full` (default): both shadow and highlight
//!   - `Outer`: shadow and highlight outside the source boundary
//!   - `Inner`: shadow and highlight inside the source boundary
//!
//! `scratch` must contain three render targets of the same dimensions as `dest`.
//! This function allocates nothing itself.

use flighthq_filters::BevelFilter;
use flighthq_filters::BevelType;

use crate::blur_filter::apply_box_blur_filter_to_gl;
use crate::filter_pass::clear_gl_render_target;
use crate::tint_shader::{apply_gl_blit_offset_pass, apply_gl_blit_pass, apply_gl_tint_pass};
use crate::{GlRenderState, GlRenderTarget};

/// Applies a bevel filter to `source`, writing the result to `dest`.
pub fn apply_bevel_filter_to_gl(
    state: &GlRenderState,
    source: &GlRenderTarget,
    dest: &GlRenderTarget,
    scratch: &[&GlRenderTarget; 3],
    filter: &BevelFilter,
) {
    let angle = filter.angle.unwrap_or(45.0).to_radians();
    let distance = filter.distance.unwrap_or(4.0);
    let dx = angle.cos() * distance;
    let dy = angle.sin() * distance;
    let shadow_color = filter.shadow_color.unwrap_or(0x000000);
    let shadow_alpha = filter.shadow_alpha.unwrap_or(1.0);
    let highlight_color = filter.highlight_color.unwrap_or(0xffffff);
    let highlight_alpha = filter.highlight_alpha.unwrap_or(1.0);
    let strength = filter.strength.unwrap_or(1.0);
    let quality = filter.quality.unwrap_or(1).max(1);
    let knockout = filter.knockout.unwrap_or(false);
    let bevel_type = filter.bevel_type.unwrap_or(BevelType::Full);

    let tinted = scratch[0];
    let blurred = scratch[1];
    let blur_temp = scratch[2];

    // Build the shared blur basis (neutral white tint preserves alpha shape).
    apply_gl_tint_pass(state, source, tinted, 0xffffff, 1.0, strength);
    apply_box_blur_filter_to_gl(
        state,
        tinted,
        blurred,
        blur_temp,
        filter.blur_x.unwrap_or(4.0),
        filter.blur_y.unwrap_or(4.0),
        quality,
    );

    clear_gl_render_target(state, dest);

    // `Inner` places shadow at -offset and highlight at +offset; `Full`/`Outer`
    // are the opposite. Sign of the highlight/shadow offsets flips accordingly.
    let inner = bevel_type == BevelType::Inner;
    let (shadow_dx, shadow_dy) = if inner { (-dx, -dy) } else { (dx, dy) };
    let (highlight_dx, highlight_dy) = if inner { (dx, dy) } else { (-dx, -dy) };

    apply_gl_tint_pass(state, blurred, tinted, shadow_color, shadow_alpha, 1.0);
    apply_gl_blit_offset_pass(state, tinted, dest, shadow_dx, shadow_dy);
    apply_gl_tint_pass(
        state,
        blurred,
        tinted,
        highlight_color,
        highlight_alpha,
        1.0,
    );
    apply_gl_blit_offset_pass(state, tinted, dest, highlight_dx, highlight_dy);

    if !knockout {
        apply_gl_blit_pass(state, source, dest);
    }
}

#[cfg(test)]
mod tests {}
